### ISSUE TITLE
Explorer shows different P2SH address than one supplied to API

### ISSUE THREAD

**Issue by scalahub** [ergoplatform/ergo#935]
Consider the following output of Swagger API captured in screenshot

https://gist.github.com/scalahub/ed96d7e8321dedf0c1b84ab1ac593409

The call created two boxes paying to two addresses as follows

P2SH address corresponding to "{1 < 2}" (7g5LhysK7mxX8xmZdPLtFE42wwxGFjpp8VofStb)
P2S address corresponding to "{1 < 2}" (4MQyML64GnzMxZgm)

The output was the following txid
fb8261828d8955c15d9ff6d7e1e0c3d6935228ff3c0af5effd95ce08018ecc76

which is available on the block explorer at:
https://explorer.ergoplatform.com/en/transactions/fb8261828d8955c15d9ff6d7e1e0c3d6935228ff3c0af5effd95ce08018ecc76

As can be seen, the output addresses are

5p1BnDBxe5KENAspbNnVzZEkvohygjPibgbnTHCB4QPp1dx9Y3ts96bSmhZ3E6ZAUhTebXeh54
and
MQyML64GnzMxZgm

While the P2S address is correct, the P2SH address is different. 

I would expect both to be the same as in the input to the API call, so it appears to be a bug.

**Comment by jasondavies** [comment]
I think the difference is due to:

* `7g5LhysK7mxX8xmZdPLtFE42wwxGFjpp8VofStb`: This is the short encoding of the P2SH address, i.e. it contains just the 192-bit hash required to claim any box protected by the full P2SH contract.
* `5p1BnDBxe5KENAspbNnVzZEkvohygjPibgbnTHCB4QPp1dx9Y3ts96bSmhZ3E6ZAUhTebXeh54`: This is the full P2SH contract, which contains all of the logic for protecting the box.

This difference should probably be documented, or perhaps the explorer should automatically detect these situations to avoid confusion.

(Note: I believe there is a bug in P2SH contracts at the moment.)

**Comment by scalahub** [comment]
@jasondavies is the bug you mention (in P2SH contracts) documented anywhere? Is it fixed?
cc @kushti @ergomorphic

**Comment by jasondavies** [comment]
> @jasondavies is the bug you mention (in P2SH contracts) documented anywhere? Is it fixed?

I don't think it was documented anywhere as I reported it in Slack and P2SH address support was swiftly disabled as a result.

The bug was in the implementation of `Pay2SHAddress` in `ErgoAddress`: when deriving the address, the hash was incorrectly computed using the serialised ErgoTree, whereas the contract code was computing the hash using the serialised proposition.  This meant that anyone attempting to use the buggy P2SH would lose their funds, as it's not possible to spend money locked in those addresses.

It looks like the bug was quite quickly fixed [here](https://github.com/ScorexFoundation/sigmastate-interpreter/commit/2ff7342aa89d49fa70cf382263e74bc871426959).
